### DISCUSSION TITLE
Modal triggered by external event

### DISCUSSION THREAD

**Discussion by cfepdan** [Budibase/budibase#5112] [Help]
Any idea how I could trigger a modal pop by an external event?

The external event can be a webhook or a database add. Maybe there's a listener I can set up in an embed block?

This would be helpful, e.g., for catching a slack message or an email notification and prepopulating some fields on my form.

**Comment by melohagan** [comment]
Hi @cfepdan 

Unfortunately at the moment there is no way to push a data refresh or modal pop from an external webhook. 

You could have an automation that updates a row in a table on webhook trigger. Then have a form that pre-populates it's fields from that table.
This would still require a manual refresh however.

**Reply by cfepdan** [reply]
Conceivably, I could set up some javascript to click a hidden data refresh button periodically and look for changes in the (hidden) table. If I give the submit button a special label, that should make it easy to get a handle for it.

 "Inelegant" is a kind term, but I think I could make it work until there's a better way.

**Reply by melohagan** [reply]
If you get this working, I would love to see your solution! :)

**Reply by cfepdan** [reply]
Oh I see I may have some trouble putting a script in an embed... https://github.com/Budibase/budibase/discussions/1892#discussioncomment-948759

**Reply by cfepdan** [reply]
This... works. 🤮 But I wouldn't recommend it. It uses an xss cheat to get around limitations on script tags. I have a button with text "Refresh" followed by an embed block with...

```
<div> <img id='pgm' data="dontstop" src="0" onerror = "javascript:{   
function sleep(ms) {
    return new Promise(resolve => setTimeout(resolve, ms));
}

async function refresh() {
    var data;
    var xpath = '//button[contains(.,\'Refresh\')]';
    var matchingElement = document.evaluate(xpath, document, null, XPathResult.FIRST_ORDERED_NODE_TYPE, null).singleNodeValue;
    do {
        data = document.getElementById('pgm').getAttribute('data');        
        matchingElement.click();
        await sleep(5000);
    } while (data == 'dontstop');
}
refresh();
}
">

<button style="width: 100px; height: 100px;; background-color: red;" onclick = "javascript: {
document.getElementById('pgm').setAttribute('data','pleasestop');
}
">  </div>
```

**Reply by melohagan** [reply]
So this effectively refreshes the data provider every 5 seconds then? 

There's an issue raised here to be able to do this auto-refreshing in the future: https://github.com/Budibase/budibase/issues/3992

But well done for finding a way! :)

**Reply by cfepdan** [reply]
Correct. It refreshes every 5 seconds. It actually activates in the builder interface, simulating a button push during the building process. Can be supremely annoying. But if you need to get the job done, this gets the job done as of v1.0.90.

I consider this approach to be a loophole that will probably be closed by the dev team someday, so keep that in mind please.
